keps/sig-cluster-lifecycle/wgs/1815-componentconfig-objectmeta/README.md

```diff
@@ -0,0 +1,689 @@
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+<!--
+This section is incredibly important for producing high quality user-focused
+documentation such as release notes or a development roadmap.  It should be
+possible to collect this information before implementation begins in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself.  KEP editors, SIG Docs, and SIG PM
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+-->
+
+The proposal is for adding ObjectMeta to the ComponentConfig API's to follow the standard pattern for other Kubernetes API's. This will allow for better interpolation with standard tooling like kustomize for managing config templating and substitution.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this KEP.  Describe why the change is important and the benefits to users.  The
+motivation section can optionally provide links to [experience reports][] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+The initial implementation of the ComponentConfig API's didn't include OjectMeta fields. The Kubernetes project tooling relies on using Kustomize with manually added ObjectMeta fields in the ComponentConfig manifests. This functionality was broken as strict decoders were implemented for ComponentConfig API's and forced a leninent path to be added for the current ComponentConfig API versions. Adding ObjectMeta will allow for valid usage with Kustomize and strict decoders for current and future ComponentConfig API's.
```

Comment:
Do we have any issues in k/k we can link to in this section to document the breakage?

/cc @rosti @neolit123

Reply:
Definitely link the related conversation on kubeadm: https://github.com/kubernetes/kubeadm/issues/1727

Reply:
i think we first saw the problem in kind, but i cannot seem to find our discussions around that.
possibly @BenTheElder remembers the ticket names better.

the summary was:
- kind uses kubeadm
- kind used to use kustomize to patch the kubeadm component config
- kubeadm has custom strict unmarshaling (warnings only) and has no "name" field in its component config.
- kind had to manually add "name" field to the kubeadm component config to conform with kustomize
- kubeadm always printed a warning for the unknown "name" field during "init/join" etc.

Reply:
Yeah, I haven't been able to find a related issue in the Kind repo.

Reply:
Added the kubeadm issue

Reply:
I can't find it with a quick search either but that sounds about right.
we moved away from kustomize for this particular problem for other reasons, it wasn't a particulary big deal to inject metadata.name, patch, then strip it out.

Reply:
I'll mark this resolved for now since I've added the kubeadm issue link. We can always link to more issues down the road if we find them.